massa-consensus-worker/src/consensus_worker.rs

```diff
@@ -34,8 +41,10 @@ pub struct ConsensusWorker {
     latest_final_periods: Vec<u64>,
     /// clock compensation
     clock_compensation: i64,
-    /// Final block stats `(time, creator)`
-    final_block_stats: VecDeque<(MassaTime, Address)>,
+    /// Final block stats `(time, creator, is_from_protocol)`
+    final_block_stats: VecDeque<(MassaTime, Address, bool)>,
+    /// Blocks that come from protocol used for stats and ids are removed when inserted in `final_block_stats`
```

Comment:
What if the block never becomes final because it ends up in a stale clique and is discarded for example ? Does it stay in this container ?
Maybe a good way to prevent that would be to avoid this HashSet and have an "origin" property for blocks in graph, alongside the `storage` property that follows blocks through their lifecycle in graph

Reply:
It can be a big overload just for desynchronization to have one more parameter with origin everywhere maybe we could just add a prune for this hashset